### ISSUE TITLE
private submodule checkout fails @v2

### ISSUE THREAD

**Issue by tvainika** [actions/checkout#116]
I've private git submodule, which contains our shared i18n resouces, with same organization owner as my main repo

I've configured git submodule in main repository as
```
$ cat .gitmodules
[submodule "xxx-i18n"]
	path = xxx-i18n
	url = ../xxx-i18n
	branch = master
```

Then I added this one from [README](https://github.com/actions/checkout/tree/c85684db76ba6ef08713c10cf4befe3318887415#Checkout-submodules) to my github action workflow.
```
name: CI

on: [push]

jobs:
  build:

    runs-on: ubuntu-latest

    steps:
    - uses: actions/checkout@v2
    - name: Checkout submodules
      shell: bash
      run: |
        auth_header="$(git config --local --get http.https://github.com/.extraheader)"
        git submodule sync --recursive
        git -c "http.extraheader=$auth_header" -c protocol.version=2 submodule update --init --force --recursive --depth=1
```

I expected this to work.

Instead I got
```
Checkout submodules
2s
##[error]Process completed with exit code 1.
Run auth_header="$(git config --local --get http.https://github.com/.extraheader)"
  auth_header="$(git config --local --get http.https://github.com/.extraheader)"
  git submodule sync --recursive
  git -c "http.extraheader=$auth_header" -c protocol.version=2 submodule update --init --force --recursive --depth=1
  shell: /bin/bash --noprofile --norc -e -o pipefail {0}
Submodule 'xxx-i18n' (https://github.com/myprivateorg/xxx-i18n) registered for path 'xxx-i18n'
Cloning into '/home/runner/work/my-main-repo/my-main-repo/xxx-i18n'...
remote: Repository not found.
fatal: repository 'https://github.com/myprivateorg/xxx-i18n/' not found
fatal: clone of 'https://github.com/myprivateorg/xxx-i18n' into submodule path '/home/runner/work/my-main-repo/my-main-repo/xxx-i18n' failed
Failed to clone 'xxx-i18n'. Retry scheduled
Cloning into '/home/runner/work/my-main-repo/my-main-repo/xxx-i18n'...
remote: Repository not found.
fatal: repository 'https://github.com/myprivateorg/xxx-i18n/' not found
fatal: clone of 'https://github.com/myprivateorg/xxx-i18n' into submodule path '/home/runner/work/my-main-repo/my-main-repo/xxx-i18n' failed
Failed to clone 'xxx-i18n' a second time, aborting
##[error]Process completed with exit code 1.
```

**Comment by eberkund** [comment]
Sub-modules support was removed in v2, I'm not sure why though

https://github.com/actions/checkout/releases/tag/v2.0.0

**Comment by tvainika** [comment]
> Sub-modules support was removed in v2, I'm not sure why though

I understood that it refers to another syntax `with: submodules` as the README I linked gives this new syntax I used.

**Comment by neutrinog** [comment]
I also need to have sub-module support. When I use v2, there's no `.git` directory so I cannot perform any many commands like checking out sub-modules.

**Comment by rodrigorn** [comment]
I have the exact same issue but I just made it work using the [Checkout multiple repos (private)](https://github.com/actions/checkout#checkout-multiple-repos-private) example
```
    - name: Checkout submodule
      uses: actions/checkout@v2.0.0
      with:
        token: '${{ secrets.GITHUB_PAT }}'
        repository: myOrg/myRepo
        path: path/to/submodule
        ref: 'master'
```
This is working just as a workaround and might not be practical for a project with a lot of submodules, but I hope it helps until #72 is resolved

**Comment by ashwinvis** [comment]
@rodrigorn It works for some use cases. Even with a single submodule, it can become inconsistent when the master branch `myOrg/myRepo` adds commits which are not tracked by the main repo.

**Comment by jleni** [comment]
This is actually blocking us from using GitHub actions. I hope it can be prioritized..

**Comment by ericsciple** [comment]
@tvainika the default token embedded in the main repo does not have access to fetch other private repositories. Instead if you supply `token: ${{ secrets.MY_GITHUB_PAT }}` then it should work.

**Comment by ericsciple** [comment]
@neutrinog no .git folder indicates git 2.18 or higher is not in your PATH. I updated the readme earlier today to make that more clear.

**Comment by jleni** [comment]
> I have the exact same issue but I just made it work using the [Checkout multiple repos (private)](https://github.com/actions/checkout#checkout-multiple-repos-private) example
> 
> ```
>     - name: Checkout submodule
>       uses: actions/checkout@v2.0.0
>       with:
>         token: '${{ secrets.GITHUB_PAT }}'
>         repository: myOrg/myRepo
>         path: path/to/submodule
>         ref: 'master'
> ```
> 
> This is working just as a workaround and might not be practical for a project with a lot of submodules, but I hope it helps until #72 is resolved

@ericsciplec @rodrigorn  is there a better alternative? I would like to avoid maintaining submodule hashes/branches in two different places. This is very error prone..

**Comment by m-hilgendorf** [comment]
I'm having the same issue, came here from the readme where it has a section on "Checkout Submodules" that does not work at all for me (I can't checkout any submodule, public or private).

**Comment by rodrigorn** [comment]
> @ericsciplec @rodrigorn is there a better alternative? I would like to avoid maintaining submodule hashes/branches in two different places. This is very error prone..
I agree completely but haven't found any better solution yet

**Comment by Lauszus** [comment]
FYI this is what I have been using:

```yml
- name: Checkout submodule
  run: |
    git config --file=.gitmodules submodule.lib/YOUR_SUBMODULE.url https://${{ secrets.CI_TOKEN }}:${{ secrets.CI_TOKEN }}@github.com/ORG/YOUR_SUBMODULE.git
    git submodule sync
    git submodule update --init --recursive
```

**Comment by ashwinvis** [comment]
@Lauszus is the CI_TOKEN a [personal access token](https://help.github.com/en/github/authenticating-to-github/creating-a-personal-access-token-for-the-command-line)?

**Comment by Lauszus** [comment]
@ashwinvis yes it is. I have updated my comment to make it more clear.

**Comment by jleni** [comment]
mm.. I am tempted to make a python script that does what @Lauszus suggests automatically for for all repositories defined in .gitmodules
@ericsciple what is the timeline for a real fix/improvement to the action?

**Comment by Lauszus** [comment]
@jleni here's a version that will checkout all repositories :)

```yml
- name: Checkout submodules using PTA token
  run: |
    git config --file .gitmodules --get-regexp url | while read url; do
      git config --file=.gitmodules $(echo "$url" | sed -E "s/git@github.com:|https:\/\/github.com\//https:\/\/${{ secrets.CI_PAT_TOKEN }}:${{ secrets.CI_PAT_TOKEN }}@github.com\//")
    done
    git submodule sync
    git submodule update --init --recursive
```

**Comment by jleni** [comment]
oh wow! that's awesome!! thanks!!

**Comment by samkit-jain** [comment]
> @tvainika the default token embedded in the main repo does not have access to fetch other private repositories. Instead if you supply `token: ${{ secrets.MY_GITHUB_PAT }}` then it should work.

The `url` in `.gitmodules` needs to be the https one for this to work.

Example `.gitmodules`
```
[submodule "module"]
	path = module
	url = https://github.com/example/module
```

Example `action.yml`
```yaml
name: Test

on: [push]

jobs:
  build:
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v2
        with:
          token: ${{ secrets.MY_GITHUB_PAT }}

      - name: Checkout submodules
        shell: bash
        run: |
          auth_header="$(git config --local --get http.https://github.com/.extraheader)"
          git submodule sync --recursive
          git -c "http.extraheader=$auth_header" -c protocol.version=2 submodule update --init --force --recursive --depth=1
```

**Comment by Lauszus** [comment]
@samkit-jain yeah, that is why I ended up using my code above: https://github.com/actions/checkout/issues/116#issuecomment-573880976, as my private repositories use `ssh` and I do not want to change all my submodules just because of the CI.

**Comment by ericsciple** [comment]
@Lauszus +1 good point. i think adding inputs for ssh support + submodule support would create a generally frictionless experience for folks looking to checkout submodules.

**Comment by daveespo** [comment]
As an alternative to @Lauszus solution to rewrite the `.gitmodules` file with the HTTP authentication info and a PAT, you could use this Action

https://github.com/webfactory/ssh-agent

You can generate an SSH keypair and add the public key to the repository you need access to (a normal Deploy Key) and then add the private key as a secret in your repo that's running the GH Action

I'm not suggesting this is great, but until there's native support for at least something resembling same-org mutual authentication for actions, this may be as good as it gets.

**Comment by alicia** [comment]
Similar to some of the comments above, I got it working with the following, where GITHUB_ACCESS_TOKEN is a personal access token in the format username:token, that is base64 encoded. Would be great submodules and private submodules to be supported more directly. 

```
git config --global url."https://github.com/".insteadOf "git@github.com:"
git submodule sync --recursive
git -c "http.extraheader=Authorization: basic ${GITHUB_ACCESS_TOKEN}" -c protocol.version=2 submodule update --init --force --recursive --depth=1
```

**Comment by jwsi** [comment]
I got frustrated switching my submodules from SSH to HTTPS and dealing with PATs so I decided to build my own action for those wishing to checkout private or public submodules via SSH in their workflows: [submodule-checkout](https://github.com/jwsi/submodule-checkout)

**Comment by r-k-b** [comment]
thanks to @alicia 's work, we also got private ssh submodules working! However, we ended up with a slightly different result:

```yaml
on: push
jobs:
  check-elm:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v2
    - name: Checkout submodules
      shell: bash
      run: |
        # From https://github.com/actions/checkout/issues/116#issuecomment-583221947
        git config --global url."https://github.com/".insteadOf "git@github.com:"
        git submodule sync --recursive
        git -c "http.extraheader=Authorization: basic ${{secrets.GITHUB_ACCESS_TOKEN}}" -c protocol.version=2 submodule update --init --force --recursive --depth=1
    - uses: actions/setup-node@v1
      with:
        node-version: '8.16.0'
    - run: npm run test
```

The main difference being `${{secrets.GITHUB_ACCESS_TOKEN}}` rather than `${GITHUB_ACCESS_TOKEN}`.

**Comment by chrispat** [comment]
Please see https://github.com/actions/checkout/issues/81#issuecomment-585935021 for some background on why we chose to not implement support for submodules in the initial version of the v2 action.

And keep an eye on the pull requests for a document discussing what features we are looking at adding.

**Comment by davidhesselbomvcc** [comment]
I tried

```yaml
    steps:
    - uses: actions/checkout@v2
      with:
        token: ${{ secrets.RUNNER_TOKEN }}
        lfs: true
    - name: Checkout submodules
      shell: bash
      run: |
        auth_header="$(git config --local --get http.https://github.com/.extraheader)"
        git submodule sync --recursive
        git -c "http.extraheader=$auth_header" -c protocol.version=2 submodule update --init --force --recursive --depth=1
```
but it fails due to LFS stuff:

> Downloading [file in the submodule]
Error downloading object: file in the submodule: Smudge error: Error downloading file in the submodule: LFS: Client error: https://github-cloud.s3.amazonaws.com/[a very long URL]

Any ideas?

**Comment by 766F6964** [comment]
> @jleni here's a version that will checkout all repositories defined in `.gitmodules` :)
> 
> It works for both `ssh` and `https`:
> 
> ```yaml
> - name: Checkout submodules using a PAT
>   run: |
>     git config --file .gitmodules --get-regexp url | while read url; do
>       git config --file=.gitmodules $(echo "$url" | sed -E "s/git@github.com:|https:\/\/github.com\//https:\/\/${{ secrets.CI_PAT }}:${{ secrets.CI_PAT }}@github.com\//")
>     done
>     git submodule sync
>     git submodule update --init --recursive
> ```

This solution seems to work perfectly for linux runners. 
However, windows runners seem to have problems with the formatting the syntax.
I am being prompted with the following error:


```
Run git config --file .gitmodules --get-regexp url | while read url; do
At D:\a\_temp\c3191bea-0f4b-4580-bbb4-1ddf2bbfd947.ps1:2 char:68
+ git config --file .gitmodules --get-regexp url | while read url; do
+                                                                    ~
Missing statement body in do loop.
+ CategoryInfo          : ParserError: (:) [], ParseException
+ FullyQualifiedErrorId : MissingLoopStatement
```
Is this a problem with powershell formatting the commands incorrectly? 
Do you have a workaround that works for windows runners too?

**Comment by ashwinvis** [comment]
@766F6964 it does not work because it uses piping, `while` loop and and `read` which are available in bash. You should try https://github.com/actions/checkout/issues/116#issuecomment-583221947 instead.

**Comment by 766F6964** [comment]
> @766F6964 it does not work because it uses piping, `while` loop and and `read` which are available in bash. You should try [#116 (comment)](https://github.com/actions/checkout/issues/116#issuecomment-583221947) instead.

I was not able to get this approach working for a windows runner either.

**Comment by ericsciple** [comment]
A submodule input has been added now. I merged to master, try it out and let me know. Collecting feedback and will update the `v2` tag next week.

Also I'm hoping to merge support for SSH soon.

**Comment by 766F6964** [comment]
@ericsciple What tag can we use for now to test this?

**Comment by ericsciple** [comment]
`master`

**Comment by m-hilgendorf** [comment]
@ericsciple the most recent commit causes the error: 

```
git@github.com: Permission denied (publickey).
```

for all submodules (public or private) when using github's hosted runner. This is true whether you set `submodules: true` or `submodules: recursive`. 

I recreated this in the following repo:

https://github.com/m-hilgendorf/actions-submodules

**Comment by ericsciple** [comment]
@m-hilgendorf SSH support not added yet. Will add tomorrow.

You can probably use the hack `git config --global url."https://github.com/".insteadOf "git@github.com:"` (run before checkout)

Interesting question whether the checkout action should add those flags.

**Comment by ericsciple** [comment]
I created a new [issue](https://github.com/actions/checkout/issues/176) to collect feedback on the `submodules` input - in master now, will update `v2` tag next week.

**Comment by ericsciple** [comment]
Also updated master with a new input `ssh-key` or when not using ssh, adds `url.https://github.com/.insteadOf git@github.com:`

**Comment by daveespo** [comment]
@ericsciple  -- just tried the new ssh-key param but no joy .. (the SSH private key that I'm passing in via Secret is one that works with the ssh-agent hack I described earlier in this thread)

```
2020-03-11T20:08:31.0562996Z Current runner version: '2.165.2'
2020-03-11T20:08:31.0564725Z Prepare workflow directory
2020-03-11T20:08:31.0791957Z Prepare all required actions
2020-03-11T20:08:31.0806683Z Download action repository 'actions/checkout@master'
2020-03-11T20:08:33.0116518Z Download action repository 'sfdx-actions/setup-sfdx@v1'
2020-03-11T20:08:33.2496238Z Download action repository 'nanasess/setup-chromedriver@master'
2020-03-11T20:08:40.5559823Z Download action repository 'actions/upload-artifact@v1'
2020-03-11T20:08:41.1828303Z ##[group]Run actions/checkout@master
2020-03-11T20:08:41.1828907Z with:
2020-03-11T20:08:41.1829353Z   lfs: true
2020-03-11T20:08:41.1831532Z   ssh-key: ***
2020-03-11T20:08:41.1831924Z   submodules: true
2020-03-11T20:08:41.1832300Z   repository: <redacted>
2020-03-11T20:08:41.1832778Z   token: ***
2020-03-11T20:08:41.1833150Z   ssh-strict: true
2020-03-11T20:08:41.1833521Z   persist-credentials: true
2020-03-11T20:08:41.1833910Z   clean: true
2020-03-11T20:08:41.1834281Z   fetch-depth: 1
2020-03-11T20:08:41.1834679Z ##[endgroup]
2020-03-11T20:08:41.6235786Z Added matchers: 'checkout-git'. Problem matchers scan action output for known warning or error strings and report these inline.
2020-03-11T20:08:41.6237082Z Syncing repository: <redacted>/patronsignup
2020-03-11T20:08:41.6237958Z Working directory is '/home/runner/work/patronsignup/patronsignup'
2020-03-11T20:08:41.6310463Z [command]/usr/bin/git version
2020-03-11T20:08:41.6458784Z git version 2.25.1
2020-03-11T20:08:41.6491554Z [command]/usr/bin/git lfs version
2020-03-11T20:08:41.7979462Z git-lfs/2.10.0 (GitHub; linux amd64; go 1.13.4)
2020-03-11T20:08:41.8001760Z Deleting the contents of '/home/runner/work/patronsignup/patronsignup'
2020-03-11T20:08:41.8013376Z [command]/usr/bin/git init /home/runner/work/patronsignup/patronsignup
2020-03-11T20:08:41.8095844Z Initialized empty Git repository in /home/runner/work/patronsignup/patronsignup/.git/
2020-03-11T20:08:41.8104612Z [command]/usr/bin/git remote add origin git@github.com:<redacted>/patronsignup.git
2020-03-11T20:08:41.8155136Z [command]/usr/bin/git config --local gc.auto 0
2020-03-11T20:08:41.8200346Z [command]/usr/bin/git config --local --name-only --get-regexp core\.sshCommand
2020-03-11T20:08:41.8240012Z [command]/usr/bin/git submodule foreach --recursive git config --local --name-only --get-regexp 'core\.sshCommand' && git config --local --unset-all 'core.sshCommand' || :
2020-03-11T20:08:41.8606381Z [command]/usr/bin/git config --local --name-only --get-regexp http\.https\:\/\/github\.com\/\.extraheader
2020-03-11T20:08:41.8641874Z [command]/usr/bin/git submodule foreach --recursive git config --local --name-only --get-regexp 'http\.https\:\/\/github\.com\/\.extraheader' && git config --local --unset-all 'http.https://github.com/.extraheader' || :
2020-03-11T20:08:41.8916066Z Temporarily overriding GIT_SSH_COMMAND="/usr/bin/ssh" -i "$RUNNER_TEMP/3e0953a3-c730-442d-9693-4efc5a4164fb" -o StrictHostKeyChecking=yes -o CheckHostIP=no -o "UserKnownHostsFile=$RUNNER_TEMP/3e0953a3-c730-442d-9693-4efc5a4164fb_known_hosts"
2020-03-11T20:08:41.8921024Z [command]/usr/bin/git config --local core.sshCommand "/usr/bin/ssh" -i "$RUNNER_TEMP/3e0953a3-c730-442d-9693-4efc5a4164fb" -o StrictHostKeyChecking=yes -o CheckHostIP=no -o "UserKnownHostsFile=$RUNNER_TEMP/3e0953a3-c730-442d-9693-4efc5a4164fb_known_hosts"
2020-03-11T20:08:41.8960017Z [command]/usr/bin/git config --local http.https://github.com/.extraheader AUTHORIZATION: basic ***
2020-03-11T20:08:41.9006512Z [command]/usr/bin/git lfs install --local
2020-03-11T20:08:41.9487647Z Updated git hooks.
2020-03-11T20:08:41.9488142Z Git LFS initialized.
2020-03-11T20:08:41.9510631Z [command]/usr/bin/git -c protocol.version=2 fetch --no-tags --prune --progress --no-recurse-submodules --depth=1 origin +c485eed0d330b13b19d081fb186d2f35836986ae:refs/remotes/origin/master
2020-03-11T20:08:42.3528109Z ERROR: Repository not found.
2020-03-11T20:08:42.3535717Z ##[error]fatal: Could not read from remote repository.
2020-03-11T20:08:42.3539415Z 
2020-03-11T20:08:42.3540237Z Please make sure you have the correct access rights
2020-03-11T20:08:42.3540662Z and the repository exists.
2020-03-11T20:08:42.3541669Z The process '/usr/bin/git' failed with exit code 128
```

**Comment by daveespo** [comment]
And my workflow YAML looks like
```
    steps:
      - uses: actions/checkout@master
        with:
          lfs: true
          ssh-key: ${{ secrets.SFDX_COMMON_SSH_KEY }}
          submodules: true
```

**Comment by ericsciple** [comment]
@daveespo thanks for trying it out! Let's troubleshoot on a new issue. Curious if you're using relative or absolute submodule URLs?

**Comment by jef** [comment]
> An input `submodules` has been added now. I merged to master, try it out and let me know. Collecting feedback and will update the `v2` tag next week.
> 
> Also I'm hoping to merge support for SSH soon.

This is awesome! I have been using with our org's private actions. Works great! Using via HTTPS and token.

**Comment by ziocleto** [comment]
> I created a new [issue](https://github.com/actions/checkout/issues/176) to collect feedback on the `submodules` input - in master now, will update `v2` tag next week.

I can give my feekback it's all good now on our end use @master, although we have an interesting case in which for some reason adding a fresh new submodules and using v2 won't work while it did work before with a submodule that was located on the root folder of the repo hmmm, anway, all good now! Every submodule is working, please let us know when you'll have a tag ready so we can remove the dangeour master tag from CD/CI!

**Comment by 766F6964** [comment]
@ericsciple Any ETA when the submodules checkout will be merged, and become available under the v2 tag?

**Comment by ericsciple** [comment]
Thanks all for the feedback! I updated V2 to include the recent submodule/SSH changes from master.

**Comment by bitgandtter** [comment]
im still facing `/usr/bin/git config --global url.https://github.com/.insteadOf git@github.com:` why is this forced? if we have our submodules configured on .gitmodules and use ssh why change it. Not sure if options are ok or doc not miss to explain it.

My config:

```
- name: Checkout
        uses: actions/checkout@v2
        with:
          ssh-key: ${{ secrets.GH_SSH_KEY }}
          submodules: recursive
```

**Comment by vadz** [comment]
FWIW this still doesn't seem to work for me neither with a private repository, using just
```
      - uses: actions/checkout@v2
        submodules: true
```
The log just says
```
##[error]fatal: repository 'https://github.com/xxx/yyy.git/' not found
```
but using the same URL with a test PAN from the command line works, so I have no idea what could be wrong here.

I've also tried using the master version, but with exactly the same result.

**Comment by ashwinvis** [comment]
What worked: create new personal access token (PAT)
- Generate PAT: https://github.com/settings/tokens
- And add the secret: https://github.com/<-- username -->/<-- repo -->/settings/secrets/new, give it a unique name say `MY_REPO_PAT`.
- Configure action as follows:
```yaml
    steps:
    - name: Checkout
      uses: actions/checkout@v2
      with:
        token: ${{ secrets.MY_REPO_PAT }}
        submodules: recursive
```

**Comment by vadz** [comment]
Thanks, this is what we've ended up by doing, but it just seems strange that such a simple operation (fully checking out the repository with its submodules) doesn't work out of the box for private repositories. IMO it should at least be clearly mentioned in the README as I lost quite some time trying to understand why it didn't work.

**Comment by ashwinvis** [comment]
I second that, this is really unclear in the README, I lost a lot of time debugging and scouring through issue pages too :cry:.

**Comment by vadz** [comment]
BTW, sorry if this is off-topic here, but what kind of tokens do people use in this case? AFAICS tokens are always associated with an account, i.e. give access to all the repositories the person has access to, which is not ideal, as it implies that the CI scripts of any of the private organizations I work with have access to all my other repositories, including those of the different private organizations.

We also looked at the deploy keys, but those can't be reused for the different repositories and we have recursive submodules in our submodule, so maintaining them all and using the correct key for each checkout would be a lot of fun that we'd rather avoid.

Is creating a specialized "CI" pseudo-user in each organization the only solution to be able to easily check out a private repository with its submodules?

**Comment by Lauszus** [comment]
@vadz I've personally set up an extra account in our organization, so it has access to all the private repositories, but without having access to my personal repositories.

I also use this account for different IoT devices to be able to access Github via ssh.

**Comment by baudehlo** [comment]
In case anyone comes here and none of the suggestions work, I finally figured out what caused this breakage for me: When you enable Actions it silently creates a PAT in your account. But if you have SAML/SSO enabled, while the token works for the main account, it fails on submodules. That is until you go into Settings -> Developer -> PAT and enable SSO for the token.

Nothing I found mentioned this when search for the error messages in question.

**Comment by chrispat** [comment]
Actions does not create a PAT in your account.  The token we generate for each job is a GitHub App per repository token and can only read the repo that the workflow is defined in.

**Comment by vadz** [comment]
Thanks @chrispat for clarifying this, I was wondering why we didn't see this automatically created PAT in our organization. But as long as I'm writing this, I'd also like to say that, if you're taking enhancement suggestions, it would be great if it were (optionally) possible to generate these per-job tokens with access to all organization repositories and not just the current one too, as it's still very surprising that something as basic as checking out a repository with its submodules doesn't work out of the box when using actions.

**Comment by baudehlo** [comment]
Odd, I don't recall creating the token. It must have been made for one of the alternate solutions above. In which case the same applies - you need to click Enable SSO on the PAT to allow the solution here: https://github.com/actions/checkout/issues/116#issuecomment-573880976 to work.

And thus private submodules still don't "just work" with this action.

**Comment by kodybrown** [comment]
Just a side-note, that I tried several of these options and couldn't get them working.. But, it turns out I didn't include the correct permissions in my PAT. Just a reminder to make sure you're creating it correctly! :)

**Comment by kcris** [comment]
late to the party, but seems to be working with more recent versions, e.g. 2.3.1

```
    - uses: actions/checkout@v2.3.1
      with:
        token: ${{secrets.PAT_MY_TOKEN}}
        submodules: recursive
        #lfs: true
```

**Comment by b-m-f** [comment]
Read only version:
- create SSH keypair on your computer
- add public part as deploy key to private submodule
- add private key as secret in repo where action is triggered. E.g SSH_KEY
- add Secret to env with
```
 env:
          SSH_KEY: ${{secrets.SSH_KEY}}
```
- do this before pulling the submodules in the run configuration
```
mkdir $HOME/.ssh && echo "$THEME_SSH" > $HOME/.ssh/id_rsa && chmod 600 $HOME/.ssh/id_rsa
```

**Comment by stephencweiss** [comment]
@b-m-f can you elaborate? What does your `.yml` look like?

Something like this?
```
jobs:
    build:
        runs-on: ubuntu-latest
        steps:
            - name: Fetch submodules
              shell: bash
              env:
                  SSH_KEY: ${{secrets.SSH_KEY}}
              run: |
                  mkdir $HOME/.ssh && echo "$SSH_KEY" > $HOME/.ssh/id_rsa && chmod 600 $HOME/.ssh/id_rsa 
                  git submodule update --init --recursive
```
Or are you using `actions/checkout@v2.3.1`?

**Comment by b-m-f** [comment]
> @b-m-f can you elaborate? What does your `.yml` look like?
> 
> Something like this?
> 
> ```
> jobs:
>     build:
>         runs-on: ubuntu-latest
>         steps:
>             - name: Fetch submodules
>               shell: bash
>               env:
>                   SSH_KEY: ${{secrets.SSH_KEY}}
>               run: |
>                   mkdir $HOME/.ssh && echo "$SSH_KEY" > $HOME/.ssh/id_rsa && chmod 600 $HOME/.ssh/id_rsa 
>                   git submodule update --init --recursive
> ```
> 
> Or are you using `actions/checkout@v2.3.1`?

Below is my configuration. It works smoothly. Give me a few days and I will follow up this solution with a blog post.

```
# A workflow run is made up of one or more jobs that can run sequentially or in parallel
jobs:
  build:
    runs-on: ubuntu-latest

    steps:
      # Checks-out your repository under $GITHUB_WORKSPACE, so your job can access it
      - uses: actions/checkout@v2
        with:
          repo-token: ${{ secrets.GITHUB_TOKEN }}
      - name: Trigger release build
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
          THEME_SSH: ${{secrets.THEME_SSH}}
        run: |
          mkdir $HOME/.ssh && echo "$THEME_SSH" > $HOME/.ssh/id_rsa && chmod 600 $HOME/.ssh/id_rsa && git submodule update --init --recursive && docker login docker.pkg.github.com -u $GITHUB_ACTOR --password $GITHUB_TOKEN && docker build . -t docker.pkg.github.com/b-m-f/blog
```

**Comment by stephencweiss** [comment]
thanks @b-m-f ! Excited to see the blog post! 💯

**Comment by b-m-f** [comment]
Hi @stephencweiss,

You can find the article at https://ehlers.berlin/blog/private-submodules-in-github-ci/ :)

**Comment by MattCrist** [comment]
@b-m-f Tried out the steps in your blog post but it doesn't seem to work.  Always get an error: ```git@github.com: Permission denied (publickey).```

**Comment by MattCrist** [comment]
> @b-m-f Tried out the steps in your blog post but it doesn't seem to work. Always get an error: `git@github.com: Permission denied (publickey).`

This was my fault, when I created my key I added a password.

**Comment by iamsushi10** [comment]
> @jleni here's a version that will checkout all repositories defined in `.gitmodules` :)
> 
> It works for both `ssh` and `https`:
> 
> ```yaml
> - name: Checkout submodules using a PAT
>   run: |
>     git config --file .gitmodules --get-regexp url | while read url; do
>       git config --file=.gitmodules $(echo "$url" | sed -E "s/git@github.com:|https:\/\/github.com\//https:\/\/${{ secrets.CI_PAT }}:${{ secrets.CI_PAT }}@github.com\//")
>     done
>     git submodule sync
>     git submodule update --init --recursive
> ```

This worked for me. Thanks a lot! 👍

**Comment by oberoivarun** [comment]
I keep getting the same error no matter what I try. 
```
fatal: could not read Username for 'https://github.com': No such device or address
```

**Comment by vavsab** [comment]
I have 1 repo that contains 2 private submodules.

The next workaround worked for me:
1. Create 2 SSH keys by running `ssh-keygen -t ed25519` twice.
2. Put each public key as deployment key into each private submodule repo.
3. Put 2 private keys as a secret into the main repo.
4. Change your action a bit
```yaml
      - uses: shaunco/ssh-agent@git-repo-mapping # this action will configure git to use the right SSH key per each repo. 
        with:
          ssh-private-key: |
                ${{ secrets.SUBMODULE_1_SSH_PRIVATE_KEY}}
                ${{ secrets.SUBMODULE_2_SSH_PRIVATE_KEY}}
          repo-mappings: |
            github.com/username_or_organization/submodule1_repo_name
            github.com/username_or_organization/submodule2_repo_name
      - uses: actions/checkout@v2
        with:
          submodules: recursive # this is important not to forget
```

**Comment by gpetrovic-meltin** [comment]
Using PAT as per: https://github.com/actions/checkout/issues/116#issuecomment-644419389 doesn't work for us because our SSH keys are setup using a CA.

We get this error:
`remote: This repository requires SSH certificate authentication. Contact the owner to receive a certificate.`

I tried to setup a separate clone step (instead of using submodules), with
```yaml
- name: clone other repo
        uses: actions/checkout@v2
        with:
          repository: org/other_repo
          token: ${{ secrets.MY_PAT }}
```

but that also didn't work for the same reason.

Seems we might need to generate CA-based SSH keys, which is less than ideal.
Or alternatively use git subtree?

**Comment by Lauszus** [comment]
@gpetrovic-meltin you can use the code snippet in my comment: https://github.com/actions/checkout/issues/116#issuecomment-573880976. I'm personally using that in all my private projects with submodules.

**Comment by erichfw** [comment]
For me, the issue was that the default token used by the github action didn't have access to clone the submodule repository. I provided access by creating a new [secret|https://docs.github.com/en/actions/reference/encrypted-secrets] in my private repo that contains a copy of my SSH github-private key. This private key provides access to both repositories. For more security, you could also use a more restrictive PAT and the [token a keyword.|https://github.com/actions/checkout]. Note that the key used needs to have access to both repositories cos it will be used for cloning the parent repository and the submodule repository. 

```
  steps:
    - uses: actions/checkout@v2
      with: 
        submodules: true
        ssh-key: ${{ secrets.xxx}}
```

**Comment by vavsab** [comment]
> For me, the issue was that the default token used by the github action didn't have access to clone the submodule repository. I provided access by creating a new [secret](https://docs.github.com/en/actions/reference/encrypted-secret) in my private repo that contains a copy of my SSH github-private key. This private key provides access to both repositories. For more security, you could also use a more restrictive PAT and the [token a keyword](https://github.com/actions/checkout). Note that the key used needs to have access to both repositories cos it will be used for cloning the parent repository and the submodule repository.
> 
> ```
>   steps:
>     - uses: actions/checkout@v2
>       with: 
>         submodules: true
>         ssh-key: ${{ secrets.xxx}}
> ```

Did you succeed to add the same SSH public key into 2 repos? In my case github throws an error when I try to add it to the 2nd repo.

**Comment by fbernaly** [comment]
> What worked: create new personal access token (PAT)
> 
> * Generate PAT: https://github.com/settings/tokens
> * And add the secret: https://github.com/<-- username -->/<-- repo -->/settings/secrets/new, give it a unique name say `MY_REPO_PAT`.
> * Configure action as follows:
> 
> ```yaml
>     steps:
>     - name: Checkout
>       uses: actions/checkout@v2
>       with:
>         token: ${{ secrets.MY_REPO_PAT }}
>         submodules: recursive
> ```

This worked for me!! Thanks @ashwinvis

**Comment by wnadim92** [comment]
got this `  /usr/bin/git -c protocol.version=2 submodule update --init --force --depth=1
  Error: fatal: No url found for submodule path 'themes/github-style' in .gitmodules
  Error: The process '/usr/bin/git' failed with exit code 128
`

was trying to use azure static web apps sourcing a hugo repo in my private github account, the build/release gave those errors

**Comment by TiagoGalvaoChange** [comment]
> 

@vavsab  when using shaunco/ssh-agent@git-repo-mapping
have you faced the need of pointing to a submodule using repo-mappings with the need to point to specific branch?
with .gitmodules file there is the possibility of defining branch...

**Comment by vavsab** [comment]
> > 
> 
> @vavsab when using shaunco/ssh-agent@git-repo-mapping have you faced the need of pointing to a submodule using repo-mappings with the need to point to specific branch? with .gitmodules file there is the possibility of defining branch...

@TiagoGalvaoChange no, I haven't. My .gitmodules only contain links to specific commits. So I may fork and add more logic to it to cover your case.

**Comment by olevskiy** [comment]
Maybe it will help somebody. This solution works when you want to keep flexibility of URL repos and still use GitHub Actions with Deploy Keys to access private submodules:

`
      - name: Checkout
        uses: actions/checkout@v2

      - name: Clone Submodule
        run: |
            mkdir -p $HOME/.ssh
            echo '${{ secrets.SUBMODULE_REPO_DEPLOY_KEY }}' > $HOME/.ssh/ssh.key
            chmod 600 $HOME/.ssh/ssh.key
            export GIT_SSH_COMMAND="ssh -i $HOME/.ssh/ssh.key"
            git submodule set-url <path-to-submodule> git@github.com:<organization/submodule>.git
            git submodule update --init --recursive
            git submodule set-url <path-to-submodule> https://github.com/<organization/submodule>.git
`

**Comment by sfullerbeckman** [comment]
> 



> Maybe it will help somebody. This solution works when you want to keep flexibility of URL repos and still use GitHub Actions with Deploy Keys to access private submodules:
> 
> ```
>   - name: Checkout
>     uses: actions/checkout@v2
> 
>   - name: Clone Submodule
>     run: |
>         mkdir -p $HOME/.ssh
>         echo '${{ secrets.SUBMODULE_REPO_DEPLOY_KEY }}' > $HOME/.ssh/ssh.key
>         chmod 600 $HOME/.ssh/ssh.key
>         export GIT_SSH_COMMAND="ssh -i $HOME/.ssh/ssh.key"
>         git submodule set-url <path-to-submodule> git@github.com:<organization/submodule>.git
>         git submodule update --init --recursive
>         git submodule set-url <path-to-submodule> https://github.com/<organization/submodule>.git
>         unset GIT_SSH_COMMAND
> ```

@arolus , that doesn't seem to work. The yaml actions file doesn't understand `unset` and `export`

**Comment by sfullerbeckman** [comment]
> > @jleni here's a version that will checkout all repositories defined in `.gitmodules` :)
> > It works for both `ssh` and `https`:
> > ```yaml
> > - name: Checkout submodules using a PAT
> >   run: |
> >     git config --file .gitmodules --get-regexp url | while read url; do
> >       git config --file=.gitmodules $(echo "$url" | sed -E "s/git@github.com:|https:\/\/github.com\//https:\/\/${{ secrets.CI_PAT }}:${{ secrets.CI_PAT }}@github.com\//")
> >     done
> >     git submodule sync
> >     git submodule update --init --recursive
> > ```
> 
> This worked for me. Thanks a lot! 👍

How did this work for you, @iamsushi10 ? I am getting  this error.

![image](https://user-images.githubusercontent.com/83652131/191334036-93c59599-19e9-48a1-8ba7-dbc0e9b1d050.png)

**Comment by sfullerbeckman** [comment]
> 



> @jleni here's a version that will checkout all repositories defined in `.gitmodules` :)
> 
> It works for both `ssh` and `https`:
> 
> ```yaml
> - name: Checkout submodules using a PAT
>   run: |
>     git config --file .gitmodules --get-regexp url | while read url; do
>       git config --file=.gitmodules $(echo "$url" | sed -E "s/git@github.com:|https:\/\/github.com\//https:\/\/${{ secrets.CI_PAT }}:${{ secrets.CI_PAT }}@github.com\//")
>     done
>     git submodule sync
>     git submodule update --init --recursive
> ```

I get this when trying to use your script...

```txt
Run git config --file .gitmodules --get-regexp url | while read url; do
sed: -e expression #1, char 86: unterminated `s' command
```

The script look fine though. Do you know what the issue is @Lauszus ?

**Comment by Lauszus** [comment]
> > 
> 
> > @jleni here's a version that will checkout all repositories defined in `.gitmodules` :)
> > It works for both `ssh` and `https`:
> > ```yaml
> > - name: Checkout submodules using a PAT
> >   run: |
> >     git config --file .gitmodules --get-regexp url | while read url; do
> >       git config --file=.gitmodules $(echo "$url" | sed -E "s/git@github.com:|https:\/\/github.com\//https:\/\/${{ secrets.CI_PAT }}:${{ secrets.CI_PAT }}@github.com\//")
> >     done
> >     git submodule sync
> >     git submodule update --init --recursive
> > ```
> 
> I get this error when trying to use your script...
> 
> ```
> Run git config --file .gitmodules --get-regexp url | while read url; do
> sed: -e expression #1, char 86: unterminated `s' command
> ```
> 
> The script look fine though. Do you know what the issue is @Lauszus ?

From your error output it looks like you are running Windows. I don't think my script works with Windows.

**Comment by asaf-kali** [comment]
For anyone who gets here in the future, adding to @fbernaly's comment, when creating a PAT:
1. Repository access -> Only select repositories -> Give access to **both**  main repo and submodule repo.
2. Permissions -> Repository permissions -> Give `Read-only` access to `Commit statuses`, `Contents`, and `Metadata`.

![image](https://user-images.githubusercontent.com/28392088/230789700-2f4c8bbc-6e74-4ff0-94a4-9b2f230c302d.png)

**Comment by richw-kai** [comment]
It's a little concerning that the only solution here is to create a PAT. The point of a Github Action is to centralise the deployment process for an organisation, that organisation may have multiple private repositories, for example defining graphql or protobuf contracts. They need importing into each repo during the build process as submodules. To rely on PAT for this process is a security concern as well as an availability risk. If the person who creates the PAT leaves the company, the whole company will now be unable to deploy due to a quite deep down and hidden error within this process.

What is the recommended solution for allowing an organisation to deploy code that references submodules from within their own organisation?

**Comment by charlielito** [comment]
> It's a little concerning that the only solution here is to create a PAT. The point of a Github Action is to centralise the deployment process for an organisation, that organisation may have multiple private repositories, for example defining graphql or protobuf contracts. They need importing into each repo during the build process as submodules. To rely on PAT for this process is a security concern as well as an availability risk. If the person who creates the PAT leaves the company, the whole company will now be unable to deploy due to a quite deep down and hidden error within this process.
> 
> What is the recommended solution for allowing an organisation to deploy code that references submodules from within their own organisation?

Because of that, I use GitHub App Installation access tokens which are not directly tied to a user but instead tied to the GitHub App installed in your organization. I use this [action](https://github.com/machine-learning-apps/actions-app-token/tree/master/) to use a previously configured and installed GitHub App to get a special token with more privileges (like access to other organization repos). This other [action](https://github.com/tibdex/github-app-token) also seems to do the same. Then, my action would look like:

```yaml
name: Some test
on:
  pull_request:
jobs:
  some_job:
    runs-on: ubuntu-22.04
    name: Some action
    steps:
      - name: Get token
        id: get_token
        uses: machine-learning-apps/actions-app-token@master
        with:
          APP_PEM: ${{ secrets.AUTH_APP_PEM }}
          APP_ID: ${{ secrets.AUTH_APP_ID }}
      - name: Checkout repository and submodules
        uses: actions/checkout@v3
        with:
          submodules: recursive
          token: ${{ steps.get_token.outputs.app_token }}
```

**Comment by mcolussi-bigw** [comment]
> Because of that, I use GitHub App Installation access tokens which are not directly tied to a user but instead tied to the GitHub App installed in your organization.

This worked for me and probably the best solution so far to not tie the submodule checkout permissions to a specific user. The only requirement is the GitHub App needs the read-only to the "Contents" permission.

**Comment by Cloudef** [comment]
Can't believe there's still no better way to do this

**Comment by zhu0629** [comment]
how can i checkout submodules on pull reqeust ???

```
      - name: Checkout
        if: github.actor == 'dependabot[bot]'
        uses: actions/checkout@v3
        with:
          token: ${{ github.token }}
          submodules: recursive

      - name: Checkout
        if: github.actor != 'dependabot[bot]'
        uses: actions/checkout@v3
        with:
          token: ${{ secrets.GH_PAT }}
          submodules: recursive
```

${{ github.token }} -> can't checkout submodules -> pull-request can use github.token
${{ secrets.GH_PAT }} -> can checkout submodules -> pull-request can't use secrets

**Comment by parth-hb** [comment]
I faced same issue and struggled atleast 2 days to make it work. If you're also struggling with the same, here's a reliable and tested fix using Deploy Keys.

### Steps

1. **Generate an SSH Key**
   ```bash
   ssh-keygen -t rsa -b 4096 -C "ci-access" -f submodule_deploy_key
   ```

2. **Add Public Key to Submodule Repo**
   - Go to your submodule repo (e.g., `my-org/my-redux-store`)
   - Navigate to: `Settings` → `Deploy Keys`
   - Click **"Add deploy key"**, paste the public key from `submodule_deploy_key.pub`
   - Make sure to enable **Allow write access** if required

3. **Add Private Key to Main Repo Secrets**
   - Go to your main repo (`my-org/my-app`)
   - Add a secret like `SUBMODULE_SSH_KEY`
   - Paste contents of `submodule_deploy_key` (the private key)

4. **Modify Workflow**
   Before updating submodules, add:
   ```yaml
   - name: Checkout main repo (no submodules)
     uses: actions/checkout@v4
     with:
        submodules: false
          
   - name: Setup SSH for submodules
     run: |
       mkdir -p ~/.ssh
       echo "${{ secrets.SUBMODULE_SSH_KEY }}" > ~/.ssh/id_rsa
       chmod 600 ~/.ssh/id_rsa
       ssh-keyscan github.com >> ~/.ssh/known_hosts

    - name: Init and update submodules
      run: |
       git submodule sync
       git submodule update --init --recursive

   ```

This should fix the common `Permission denied` or `Repository not found` errors when dealing with submodules.

---

**Works great when:**  
- You don’t want to use personal tokens  
- You want tight access control over a single repo  
- You prefer SSH over HTTPS

**Comment by parth-hb** [comment]
I faced same issue and struggled atleast 2 days to make it work. If you're also struggling with the same, here's a reliable and tested fix using Deploy Keys.

### Steps

1. **Generate an SSH Key**
   ```bash
   ssh-keygen -t rsa -b 4096 -C "ci-access" -f submodule_deploy_key
   ```

2. **Add Public Key to Submodule Repo**
   - Go to your submodule repo (e.g., `my-org/my-redux-store`)
   - Navigate to: `Settings` → `Deploy Keys`
   - Click **"Add deploy key"**, paste the public key from `submodule_deploy_key.pub`
   - Make sure to enable **Allow write access** if required

3. **Add Private Key to Main Repo Secrets**
   - Go to your main repo (`my-org/my-app`)
   - Add a secret like `SUBMODULE_SSH_KEY`
   - Paste contents of `submodule_deploy_key` (the private key)

4. **Modify Workflow**
   Before updating submodules, add:
   ```yaml
   - name: Checkout main repo (no submodules)
     uses: actions/checkout@v4
     with:
        submodules: false
          
   - name: Setup SSH for submodules
     run: |
       mkdir -p ~/.ssh
       echo "${{ secrets.SUBMODULE_SSH_KEY }}" > ~/.ssh/id_rsa
       chmod 600 ~/.ssh/id_rsa
       ssh-keyscan github.com >> ~/.ssh/known_hosts

    - name: Init and update submodules
      run: |
       git submodule sync
       git submodule update --init --recursive

   ```

This should fix the common `Permission denied` or `Repository not found` errors when dealing with submodules.

---

**Works great when:**  
- You don’t want to use personal tokens  
- You want tight access control over a single repo  
- You prefer SSH over HTTPS

**Comment by davidt-gh** [comment]
Why is that closed if the issue never solved?

**Comment by NicolasGonzalezFarabollini** [comment]
You can bypass this issue if you use the token inside a docker instance on github actions. Yes, not the most efficient, nor elegant. But it works.

For example, I just need to build a dependency from a private repo under same org. I make the yarn build inside the docker image, validate with token. Next, I extract the files from the docker image, and continue the build process with normal github steps.
